File: Leetcode/MajorityElement169.java
package Algorithms.Leetcode;

import java.util.HashMap;
import java.util.Set;

/**
 * Given an array of size n, find the majority element.
 * The majority element is the element that appears more than ⌊ n/2 ⌋ times.
 * You may assume that the array is non-empty and
 * the majority element always exist in the array.
 *
 * Created by dianaluca on 9/15/16.
 */

public class MajorityElement169 {
  public int majorityElement(int[] nums) {
    int N = nums.length;
    HashMap hm = new HashMap<Integer, Integer>();
    for(int i = 0; i < nums.length; i++){
      int val = hm.get(nums[i]) == null ? 0 : (Integer) hm.get(nums[i]);
      hm.put(nums[i], ++val);
    }

    Set<Integer> keys = hm.keySet();
    for(Integer key : keys){
      int val = (Integer)hm.get(key);
      if(val > N/2){
        return key;
      }
    }
    return -1;
  }
}

File: LeetcodePython/MajorityElement169.py
# Given an array of size n, find the majority element. The majority element is the element that appears more than ⌊ n/2 ⌋ times.

# class Solution(object):
#     def majorityElement(self, nums):
#         """
#         :type nums: List[int]
#         :rtype: int
#         """
#         d = dict()
#         n = len(nums)
#         for el in nums:
#             d[el] = d.get(el, 0) + 1
#
#         for i in range(n):
#             if d[nums[i]] > n//2:
#                 return nums[i]


from collections import Counter


class Solution(object):
    def majorityElement(self, nums):

        """
        :type nums: List[int]
        :rtype: int
        """
        d = Counter(nums)

        for el in nums:
            if d[el] > len(nums) // 2:
                return el




